chain/chain/src/finality.rs

```diff
@@ -0,0 +1,230 @@
+use crate::error::Error;
+use crate::{ChainStoreAccess, ChainStoreUpdate};
+use near_primitives::block::{Approval, BlockHeader, BlockHeaderInner};
+use near_primitives::hash::CryptoHash;
+use near_primitives::types::{AccountId, BlockIndex};
+use std::collections::{HashMap, HashSet};
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct FinalityGadgetQuorums {
+    pub last_quorum_pre_vote: CryptoHash,
+    pub last_quorum_pre_commit: CryptoHash,
+}
+
+#[derive(Debug)]
+pub struct ApprovalVerificationError {}
+
+pub struct FinalityGadget {}
+
+impl FinalityGadget {
+    pub fn process_approval(
+        &self,
+        me: &Option<AccountId>,
+        approval: &Approval,
+        chain_store_update: &mut ChainStoreUpdate,
+    ) {
+        // First update the statistics for the current block producer if the approval is created by us
+        let (total_weight, score) = match chain_store_update.get_block_header(&approval.parent_hash)
+        {
+            Ok(header) => {
+                let BlockHeader { inner: BlockHeaderInner { total_weight, score, .. }, .. } =
+                    header;
+                let total_weight = total_weight.clone();
+                let score = score.clone();
+                (Some(total_weight), Some(score))
+            }
+            Err(_) => (None, None),
+        };
+
+        if me.as_ref().map(|me| me == &approval.account_id).unwrap_or(false) {
+            let prev_weight = chain_store_update.largest_approved_weight().map(|x| x.clone());
+            let prev_score = chain_store_update.largest_approved_score().map(|x| x.clone());
```

Comment:
Young relying on error being only missing element here? Not super safe, prob should be expect when instead of using is_err below